### PR TITLE
Test if `yalc` can be used in the CI for the build tests

### PR DESCRIPTION
**Pull request by fulopkovacs**
Make sure that `@theatre/core` and `@theatre/studio` can be used with the most popular setups (`parcel`, `vite`, `vanilla rollup`, etc.).

**Note:** Our first approach was [this PR](https://github.com/theatre-js/theatre/pull/120) which had to be closed due to the limitations of `yarn workspaces`.

### CODE REVIEW

**Comment by fulopkovacs** [comment]
# Build setups

Here's a list of the build setups that I'm planning to test in this PR (feel free to add more):

- [x] Vite: we already use it internally
- [ ] Parcel 1 (WIP)
- [ ] Parcel 2 (WIP)
- [ ] vanilla rollup (WIP)
- [ ] create-react-app
- [ ] :x: NextJS 11 (see **UPDATES**)
- [ ] :x: NextJS 12 (see **UPDATES**)

WIP = the setup is ready, it just need to be added to the repo.

# Other items on the TODO list

- [ ] make sure that Parcel 2 does not use Babel (and especially the babel config in the monorepo's root)
- [ ] Add a trouble shooting guide to the docs for people using `rollup` without using the `@rollup/plugin-replace` to replace the `process.env.NODE_ENV` with its value during the build
- [ ] Monkey-patch the `process.env.NODE_ENV` variable if `process` is `undefined` (see the previous TODO item) in Theatre.js and print a warning

# UPDATES

- `Next.js v12` is currently not compatible with Theatre.js (check out [this conversation](https://discord.com/channels/870988717190426644/870988717190426647/961909489181818930) on our Discord's _#help_ channel). This issue is still being investigated.
- `Next.js v11` has the same problem as `v12`
- we ran into some problems with `yarn` workspaces, so decided to not use them for the build tests. Currently I'm checking out [`yalc`](https://github.com/wclr/yalc) as our next best option.

**Comment by DennisSmolek** [comment]
@fulop how are you clearing vite?
I'm guessing you aren't using the r3f extension?

I'm getting:
```
✘ [ERROR] [plugin vite:dep-pre-bundle] Missing "./examples/js/libs/stats.min" export in "three" package

    node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:37698:7:
      37698 │   throw new Error(
            ╵         ^

    at bail (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:37698:8)
    at resolve (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:37775:10)
    at resolveExports (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38441:12)
    at resolveDeepImport (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38459:31)
    at tryNodeResolve (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38232:20)
    at Context.resolveId (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:38040:28)
    at Object.resolveId (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:36609:55)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:61261:27
    at async /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:30513:34

  This error came from the "onResolve" callback registered here:

    node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:30492:18:
      30492 │             build.onResolve({ filter: /^[\w@][^:]/ }, async ({ path: id, importer, kind }) => {
            ╵                   ~~~~~~~~~

    at setup (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:30492:19)
    at handlePlugins (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:843:23)
    at Object.buildOrServe (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1137:7)
    at /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:2079:17
    at new Promise (<anonymous>)
    at Object.build (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:2078:14)
    at Object.build (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1927:51)
    at runOptimizeDeps (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:37342:34)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

  The plugin "vite:dep-pre-bundle" was triggered by this import

    node_modules/@theatre/r3f/node_modules/@react-three/drei/core/Stats.js:3:22:
      3 │ import StatsImpl from 'three/examples/js/libs/stats.min';
        ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

7:59:42 AM [vite] error while updating dependencies:
Error: Build failed with 1 error:
node_modules/vite/dist/node/chunks/dep-27bc1ab8.js:37698:7: ERROR: [plugin: vite:dep-pre-bundle] Missing "./examples/js/libs/stats.min" export in "three" package
    at failureErrorWithLog (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1603:15)
    at /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1249:28
    at runOnEndCallbacks (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1034:63)
    at buildResponseToResult (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1247:7)
    at /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:1356:14
    at /Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:666:9
    at handleIncomingPacket (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:763:9)
    at Socket.readFromStdout (/Users/Shared/sharedGit/r3f-theatre-fresh/node_modules/esbuild/lib/main.js:632:7)
    at Socket.emit (events.js:400:28)
    at addChunk (internal/streams/readable.js:293:12)

```

**Comment by fulopkovacs** [comment]
Hey, @DennisSmolek , sorry for the late response (holidays... ) :relaxed:!

The reason why `vite` and `r3f` are not included in this PR is because we're planning to build all the projects in [`packages/playground/src/shared`](https://github.com/theatre-js/theatre/tree/main/packages/playground/src/shared) using Netlify's build preview for every PR. These projects do use `vite`, and [one of them uses `r3f`](https://github.com/theatre-js/theatre/tree/main/packages/playground/src/shared/r3f-rocket). Unfortunately the Build Previews currently have a bug ([issue](https://github.com/theatre-js/theatre/tree/main/packages/playground/src/shared)), so we have to wait until the fix we submitted ([PR](https://github.com/netlify/build-image/pull/763)) gets merged.

I'm not really familiar with the error in your comment. Checked the playground project that uses the r3f extension, but it seems to be fine. Could you check it too to see if I'm missing something? Here's how you could do it:
1. Run this command in the cloned repo's root after you've installed the dependencies:
```shell
yarn playground
```
2. Visit [http://localhost:8080/shared/r3f-rocket](http://localhost:8080/shared/r3f-rocket) in the browser. (Replace the port with the the one that the  playground server is using if necessary)
3. You  should be presented with something like this:
image

I hope I was able to help you, please tell me if that's not the case :relaxed:.

File: scripts/build-tests/link-theatre-packages.mjs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+import path from 'path'
+import {colorize, getTestBuildProjects} from './utils.mjs'
+
+const root = path.normalize(path.join(__dirname, '..', '..'))
+const buildTestProjectAbsPaths = getTestBuildProjects(root)
+
+// The packages that should be linked with `yalc`
+const packagesToLink = ['@theatre/core', '@theatre/studio']
```
Comment (AriaMinaei) [inline scripts/build-tests/link-theatre-packages.mjs:12]: Any reason we're not also linking `@theatre/dataverse`?

Comment (AriaMinaei) [reply scripts/build-tests/link-theatre-packages.mjs:12]: Also, `@theatre/r3f`?

File: scripts/publish-to-yalc.mjs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import path from 'path'
+
+const root = path.normalize(path.join(__dirname, '..'))
+
+// Make sure the script runs in the root of the monorepo
+cd(root)
+
+process.env.USING_YALC = 'true'
+
+const packagesToPublish = ['theatre/core', 'theatre/studio']
```
Comment (AriaMinaei) [inline scripts/publish-to-yalc.mjs:13]: Same as the previous comment:
> Any reason we're not also linking `@theatre/dataverse`?

**Review by AriaMinaei — COMMENTED** [review]
Did a first pass at this draft PR. Looks very good!

**Comment by AriaMinaei** [comment]
Could this be a mac issue?

image

File: scripts/build-tests/link-theatre-packages.mjs (hunk 1 above)
Comment (fulopkovacs) [reply scripts/build-tests/link-theatre-packages.mjs]: They weren't necessary for the  build tests before. 

(But now I had to add them for the `create-react-app-r3f` project (053e3d09), so I'll resolve this conversation! :relaxed:)

File: scripts/publish-to-yalc.mjs (hunk 2 above)
Comment (fulopkovacs) [reply scripts/publish-to-yalc.mjs]: They weren't needed for the build_test projects, but eventually they got in: 053e3d09. I'll mark this conversation as 'resolved' too. :relaxed:

**Comment by fulopkovacs** [comment]
> Could this be a mac issue?
> 
> image

Note for the people looking at this PR: We talked with Aria and solved the issue by adding `node-gyp` to the project's dependencies (also had to run `yarn rebuild` before that if I remember correctly). This issue was macOS-specific, but houldn't come up again.

**Comment by fulopkovacs** [comment]
This PR ran into a error in the CI and after a considerable amount of debugging we decided to restart it from scratch. See the new version of the PR here: https://github.com/theatre-js/theatre/pull/165

**Pull request closed without merging**